cpp/open3d/t/io/file_format/FileSPLAT.cpp

```diff
@@ -0,0 +1,354 @@
+
+        float *position_ptr = static_cast<float *>(
+                pointcloud.GetPointPositions().GetDataPtr());
+        float *scale_ptr = static_cast<float *>(
+                pointcloud.GetPointAttr("scale").GetDataPtr());
+        float *rot_ptr = static_cast<float *>(
+                pointcloud.GetPointAttr("rot").GetDataPtr());
+        float *f_dc_ptr = static_cast<float *>(
+                pointcloud.GetPointAttr("f_dc").GetDataPtr());
+        float *opacity_ptr = static_cast<float *>(
+                pointcloud.GetPointAttr("opacity").GetDataPtr());
+        float *f_rest_ptr = static_cast<float *>(
+                pointcloud.GetPointAttr("f_rest").GetDataPtr());
+
+        // Read the data
+        int index = 0;
+        while (file.ReadData(buffer, SPLAT_GAUSSIAN_BYTE_SIZE)) {
+            uint8_t color[4];
+            uint8_t rotation[4];
+
+            // Copy the data into the vectors
+            std::memcpy(position_ptr + index * 3, buffer, 3 * sizeof(float));
+            std::memcpy(scale_ptr + index * 3, buffer + 3 * sizeof(float),
+                        3 * sizeof(float));
+            std::memcpy(color, buffer + 6 * sizeof(float), 4 * sizeof(uint8_t));
+            std::memcpy(rotation,
+                        buffer + (6 * sizeof(float)) + (4 * sizeof(uint8_t)),
+                        4 * sizeof(uint8_t));
+
+            // Calculate the f_dc and f_rest
+            float f_dc[3];
+            float f_rest[3];
+            for (int i = 0; i < 3; i++) {
+                f_dc[i] = ((color[i] / 255.0f) - 0.5) / SH_C0;
+                f_rest[i] = SH_C0;
+            }
+            std::memcpy(f_dc_ptr + index * 3, f_dc, 3 * sizeof(float));
+            std::memcpy(f_rest_ptr + index * 3, f_rest, 3 * sizeof(float));
+
+            // Calculate the opacity
+            float opacity[1];
+            if (color[3] == 0) {
+                opacity[0] = 0.0f;  // Handle division by zero
+            } else {
+                opacity[0] = -log(1 / (color[3] / 255.0f) - 1);
+            }
+            std::memcpy(opacity_ptr + index, opacity, sizeof(float));
+
+            float rot_float[4];
+            for (int i = 0; i < 4; i++) {
+                rot_float[i] = (rotation[i] / 128.0f) - 1.0f;
+            }
+            std::memcpy(rot_ptr + index * 4, rot_float, 4 * sizeof(float));
+
+            index++;
+            if (index % 1000 == 0) {
+                reporter.Update(file.CurPos());
+            }
+        }
+
+        // Report progress
+        reporter.Finish();
+        return true;
+    } catch (const std::exception &e) {
+        utility::LogWarning("Read SPLAT failed: {}", e.what());
+    }
+    return false;
+}
+
+bool WritePointCloudToSPLAT(const std::string &filename,
+                            const geometry::PointCloud &pointcloud,
+                            const open3d::io::WritePointCloudOption &params) {
+    FILE *splat_file = NULL;
```

Comment:
Use C++ `fstream` instead of C FILE. The advantage is better / faster error handling. We don't have to check the result of each fwrite. Instead we just get an exception if an error occurs. Here we are checking every number written for up to millions of Gaussians, so this will make a significant speed difference.

Reply:
Done